File: lib/musicxml/node/note.rb
module MusicXML
  module Node

    register :note do
      sattrs :duration, :staff, :stem, :type, :voice

      pnodes :notations
      snodes :pitch, :chord

      props :dot

      # convert a note to lilypond notation
      def to_lilypond
        "#{pitch.step.downcase}#{lilypond_octave}#{lilypond_duration}"
      end

      private

        def lilypond_duration
          case type
          when '16th' then 16
          when 'eighth' then 8
          when 'quarter' then 4
          when 'half' then 2
          when 'whole' then 1
          else ''
          end
        end

        def lilypond_octave
          octave = pitch.octave.to_i
          if octave > 3
            '\'' * (octave - 3)
          elsif octave < 3
            ',' * (3 - octave)
          else
            ''
          end
        end
    end
  end
end

File: lib/assets/src/note.js
MusicXML.Note = function(config) {
	this.config = config;
};

MusicXML.Note.prototype.toStaveNote = function() {
	var staveNote = new Vex.Flow.StaveNote({
		keys: [this.compiledPitch()],
		duration: this.compiledDuration()
	});
	if(this.config.dot) {
		staveNote.addDotToAll();
	}
	return staveNote;
};

MusicXML.Note.prototype.compiledDuration = function() {
	var compiled;

	switch(this.config.type) {
	case "whole":
		compiled = "w";
		break;
	case "half":
		compiled = "h";
		break;
	case "quarter":
		compiled = "q";
		break;
	case "eighth":
		compiled = "8";
		break;
	case "16th":
		compiled = "16";
		break;
	}

	if(this.config.dot) {
		compiled += "d";
	}

	return compiled;
};

MusicXML.Note.prototype.compiledPitch = function() {
	return this.config.pitch.step.toLowerCase() + "/" + this.config.pitch.octave;
};
